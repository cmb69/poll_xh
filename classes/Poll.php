<?php

/**
 * The polls.
 *
 * PHP version 5
 *
 * @category  CMSimple_XH
 * @package   Poll
 * @author    Christoph M. Becker <cmbecker69@gmx.de>
 * @copyright 2012-2014 Christoph M. Becker <http://3-magi.net/>
 * @license   http://www.gnu.org/licenses/gpl-3.0.en.html GNU GPLv3
 * @version   SVN: $Id$
 * @link      http://3-magi.net/?CMSimple_XH/Poll_XH
 */

/**
 * The polls.
 *
 * @category CMSimple_XH
 * @package  Poll
 * @author   Christoph M. Becker <cmbecker69@gmx.de>
 * @license  http://www.gnu.org/licenses/gpl-3.0.en.html GNU GPLv3
 * @link     http://3-magi.net/?CMSimple_XH/Poll_XH
 */
class Poll_Poll
{
    /**
     * The name.
     *
     * @var string
     */
    protected $name;

    /**
     * The maximum number of votes.
     *
     * @var int
     */
    protected $maxVotes;

    /**
     * The end date of the voting.
     *
     * @var int A UNIX timestamp.
     */
    protected $endDate;

    /**
     * The total number of votes so far.
     *
     * @var int
     */
    protected $totalVotes;

    /**
     * The votes as map from option name to number of votes.
     *
     * @var array<string, int>
     */
    protected $votes;

    /**
     * Returns the name.
     *
     * @return string
     */
    public function getName()
    {
        return $this->name;
    }

    /**
     * Returns the maximum number of votes.
     *
     * @return int A UNIX timestamp.
     */
    public function getMaxVotes()
    {
        return $this->maxVotes;
    }

    /**
     * Returns the end date of the voting.
     *
     * @return int
     */
    public function getEndDate()
    {
        return $this->endDate;
    }

    /**
     * Returns whether the poll has ended.
     *
     * @return bool
     */
    public function hasEnded()
    {
        return $this->getEndDate() <= time();
    }

    /**
     * Returns the total number of votes so far.
     *
     * @return int
     */
    public function getTotalVotes()
    {
        return $this->totalVotes;
    }

    /**
     * Returns the vote count of an option.
     *
     * @param string $optionName An option name.
     *
     * @return int
     */
    public function getVoteCount($optionName)
    {
        return $this->votes[$optionName];
    }

    /**
     * Returns the votes as map from option name to number of votes.
     *
     * @return array<string, int>
     */
    public function getVotes()
    {
        return $this->votes;
    }

    /**
     * Sets the name.
     *
     * @param string $name A name.
     *
     * @return void
     */
    public function setName($name)
    {
        $this->name = $name;
    }

    /**
     * Sets the maximum number of votes.
     *
     * @param int $number A number.
     *
     * @return void
     */
    public function setMaxVotes($number)
    {
        $this->maxVotes = $number;
    }

    /**
     * Returns the end date of the voting.
     *
     * @param int $timestamp A UNIX timestamp.
     *
     * @return void
     */
    public function setEndDate($timestamp)
    {
        $this->endDate = $timestamp;
    }

    /**
     * Sets the total number of votes so far.
     *
     * @param int $number A number.
     *
     * @return void
     */
    public function setTotalVotes($number)
    {
        $this->totalVotes = $number;
    }

    /**
     * Increases the total number of votes by one.
     *
     * @return void
     */
    public function increaseTotalVotes()
    {
        $this->totalVotes++;
    }

    /**
     * Sets the vote count for an option.
     *
     * @param string $optionName An option name.
     * @param int    $count      A vote count.
     *
     * @return void
     */
    public function setVoteCount($optionName, $count)
    {
        $this->votes[$optionName] = $count;
    }

    /**
     * Increases the vote count of an option by one.
     *
     * @param string $optionName An option name.
     *
     * @return void
     */
    public function increaseVoteCount($optionName)
    {
        $this->votes[$optionName]++;
    }

    /**
     * Sorts the votes descending by count.
     *
     * @return void
     */
    public function sortVotes()
    {
        arsort($this->votes);
    }
}

?>
