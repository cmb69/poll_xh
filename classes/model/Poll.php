<?php

/**
 * Copyright (c) Christoph M. Becker
 *
 * This file is part of Poll_XH.
 *
 * Poll_XH is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Poll_XH is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Poll_XH.  If not, see <http://www.gnu.org/licenses/>.
 */

namespace Poll\Model;

class Poll
{
    /** @var string */
    protected $name;

    /** @var int */
    protected $maxVotes;

    /** @var int Unix timestamp */
    protected $endDate;

    /** @var int */
    protected $totalVotes;

    /** @var array<string,int>*/
    protected $votes;

    public function getName(): string
    {
        return $this->name;
    }

    public function getMaxVotes(): int
    {
        return $this->maxVotes;
    }

    public function getEndDate(): int
    {
        return $this->endDate;
    }

    public function hasEnded(): bool
    {
        return $this->getEndDate() <= time();
    }

    public function getTotalVotes(): int
    {
        return $this->totalVotes;
    }

    public function getVoteCount(string $optionName): int
    {
        return $this->votes[$optionName];
    }

    /** @return array<string,int> */
    public function getVotes(): array
    {
        return $this->votes;
    }

    public function setName(string $name): void
    {
        $this->name = $name;
    }

    public function setMaxVotes(int $number): void
    {
        $this->maxVotes = $number;
    }

    public function setEndDate(int $timestamp): void
    {
        $this->endDate = $timestamp;
    }

    public function setTotalVotes(int $number): void
    {
        $this->totalVotes = $number;
    }

    public function increaseTotalVotes(): void
    {
        $this->totalVotes++;
    }

    public function setVoteCount(string $optionName, int $count): void
    {
        $this->votes[$optionName] = $count;
    }

    public function increaseVoteCount(string $optionName): bool
    {
        if (!isset($this->votes[$optionName])) {
            return false;
        }
        $this->votes[$optionName]++;
        return true;
    }

    public function sortVotes(): void
    {
        arsort($this->votes);
    }
}
