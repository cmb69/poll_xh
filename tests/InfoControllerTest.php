<?php

namespace Poll;

use ApprovalTests\Approvals;
use PHPUnit\Framework\TestCase;
use Plib\FakeSystemChecker;
use Plib\View;
use Poll\Model\DataService;

class InfoControllerTest extends TestCase
{
    public function testRendersPluginInfo(): void
    {
        $dataServiceStub = $this->createStub(DataService::class);
        $dataServiceStub->method("getFolder")->willReturn("./content/poll/");
        $sut = new InfoController(
            "./plugins/poll/",
            $dataServiceStub,
            new FakeSystemChecker(),
            new View("./views/", XH_includeVar("./languages/en.php", "plugin_tx")["poll"])
        );
        Approvals::verifyHtml($sut->defaultAction());
    }
}
