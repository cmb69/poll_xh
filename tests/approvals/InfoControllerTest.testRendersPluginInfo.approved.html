<h1>Poll 1.0beta3</h1>
<div class="poll_syscheck">
    <h2>System check</h2>
    <p class="xh_fail">Checking that the PHP version is at least 7.1.0 … failure</p>
    <p class="xh_fail">Checking that the CMSimple_XH version is at least 1.7.0 … failure</p>
    <p class="xh_fail">Checking that the Plib_XH version is at least 1.1 … failure</p>
    <p class="xh_warning">Checking that './plugins/poll/css' is writable … warning</p>
    <p class="xh_warning">Checking that './plugins/poll/languages' is writable … warning</p>
    <p class="xh_warning">Checking that './content/poll/' is writable … warning</p>
</div>
